### PR TITLE
Make skolem definition manager robust to definitions for already asserted skolems

### PR DESCRIPTION
**Pull request by ajreynol**
It makes the skolem definition manager more robust so that skolem definitions can be added for skolems that have already appeared in asserted literals. This was the initial motivation for the change in ordering.  As a result, fixes https://github.com/cvc5/cvc5/issues/8347 and fixes https://github.com/cvc5/cvc5-projects/issues/512.  It also optimizes this class in a few ways.

It also comments more on the change to PropEngine introduced here: https://github.com/cvc5/cvc5/pull/8301 which led to performance degradation on a set of string benchmarks of interest.

### CODE REVIEW

**Review by 4tXJ7f — APPROVED** [review]
LGTM

**Comment by 4tXJ7f** [comment]
Marking as "do not merge" until we have the performance results.

**Comment by ajreynol** [comment]
It seems the performance results are mixed / negative on SMT-LIB. I am considering reverting the change to the assertion order but keeping the changes to skolem definition manager.

File: src/prop/skolem_def_manager.cpp [hunk 1]
```diff
@@ -107,12 +98,22 @@ bool SkolemDefManager::hasSkolems(TNode n)
...
+        // We have skolems if we are a skolem that has a definition, or
+        // we are a Boolean term variable. For Boolean term variables, we do
+        // not make this test depend on whether the skolem has a definition,
+        // since that is prone to change if the Boolean term variable was
+        // introduced in a lemma prior to its definition being introduced.
+        // This is for example the case in strings reduction for Booleans,
+        // ground term purification for E-matching, etc.
+        if (ck == kind::SKOLEM)
+        {
+          d_hasSkolems[cur] = (d_skDefs.find(cur)!=d_skDefs.end());
```
Comment (4tXJ7f) [inline src/prop/skolem_def_manager.cpp:111]: Apply ClangFormat.

**Review by 4tXJ7f — APPROVED** [review]
LGTM

**Pull request merged**